source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -76,21 +109,149 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
+   * It is allowed by the protocol but not widely supported:
+   *
+   * See: https://www.ietf.org/rfc/rfc1035.txt
+   *
+   * The question section is used to carry the "question" in most queries,
+   * i.e., the parameters that define what is being asked. The section
+   * contains QDCOUNT (usually 1) entries.
+   */
+  for (const auto& query : context->queries_) {
+    // Try to resolve the query locally. If forwarding the query externally is disabled we will
+    // always attempt to resolve with the configured domains
+    if (isKnownDomain(query->name_) || !config_->forwardQueries()) {
+      // Determine whether the name is a cluster. Move on to the next query if successful
+      if (resolveViaClusters(context, *query)) {
+        continue;
+      }
+
+      // Determine whether we an answer this query with the static configuration
+      if (resolveViaConfiguredHosts(context, *query)) {
+        continue;
+      }
+    }
+    // TODO(abaptiste): resolve the query externally
+  }
+
+  if (context->answers_.empty()) {
+    return DnsLookupResponseCode::Failure;
+  }
+  return DnsLookupResponseCode::Success;
+}
+
+uint32_t DnsFilter::getDomainTTL(const absl::string_view domain) {
+  const auto& domain_ttl_config = config_->domainTtl();
+  const auto& iter = domain_ttl_config.find(domain);
+
+  uint32_t ttl;
+  if (iter == domain_ttl_config.end()) {
+    ttl = static_cast<uint32_t>(DEFAULT_RESOLVER_TTL.count());
+  } else {
+    ttl = static_cast<uint32_t>(iter->second.count());
+  }
+
+  return ttl;
+}
+
+bool DnsFilter::isKnownDomain(const absl::string_view domain_name) {
+  const auto& known_suffixes = config_->knownSuffixes();
+
+  // If we don't have a list of whitelisted domain suffixes, we will resolve the name with an
+  // external DNS server
+  if (known_suffixes.empty()) {
+    ENVOY_LOG(debug, "Known domains list is empty");
+    return false;
+  }
+
+  // TODO(abaptiste): Use a trie to find a match instead of iterating through the list
+  for (auto& suffix : known_suffixes) {
+    if (suffix->match(domain_name)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool DnsFilter::resolveViaClusters(DnsQueryContextPtr& context, const DnsQueryRecord& query) {
+  Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(query.name_);
+  if (cluster == nullptr) {
+    ENVOY_LOG(debug, "Did not find a cluster for name [{}]", query.name_);
+    return false;
+  }
+
+  // Return the address for all discovered endpoints
+  size_t discovered_endpoints = 0;
+  const uint32_t ttl = getDomainTTL(query.name_);
+  for (const auto& hostsets : cluster->prioritySet().hostSetsPerPriority()) {
+    for (const auto& host : hostsets->hosts()) {
+      ++discovered_endpoints;
+      ENVOY_LOG(debug, "using cluster host address {} for domain [{}]",
+                host->address()->ip()->addressAsString(), query.name_);
+      message_parser_.buildDnsAnswerRecord(context, query, ttl, host->address());
```

Comment:
do we use a concept of weights here? I know clusters have that notion and DNS allows it. Plumbing the 2 is the quesiton

Reply:
No, weights are not considered.   In the context of clusters, I looked at leveraging the load balancer, but was not able to get this working.  There's a context object whose origin still remains a mystery 👻 

If you are talking about adding weights to the addresses returned, that's will likely be a future addition.

Reply:
sure. a TODO would help

Reply:
Ok, I'll add a note